src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- and removing all edges inside one part.
+-- Complexity: /O(m log n)/.
+--
+-- @
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.empty'))                                         == Map.'Map.empty'
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.edge' (Left 1) (Right 2)))                       == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.edge' (Left 1) (Left 2)))                        == Map.'Map.empty'
+-- 'rightAdjacencyMap' (fromGraph ('Algebra.Graph.edges' [(Left 1, Right 1), (Left 1, Right 2)])) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+-- @
+fromGraph :: (Ord a, Ord b) => G.Graph (Either a b) -> AdjacencyMap a b
+fromGraph = toBipartite . (G.foldg AM.empty AM.vertex AM.overlay AM.connect)
+
+-- | Constructs the /empty graph/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' empty  == Map.'Map.empty'
+-- 'rightAdjacencyMap' empty == Map.'Map.empty'
+-- 'hasVertex' x empty       == False
+-- @
+empty :: AdjacencyMap a b
+empty = BAM Map.empty Map.empty
+
+-- | Constructs the bipartite graph comprising /a single isolated vertex/ in
+-- the left part.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (leftVertex x)  == Map.'Map.singleton' x Set.'Set.empty'
+-- 'rightAdjacencyMap' (leftVertex x) == Map.'Map.empty'
+-- 'hasEdge' x y (leftVertex x)       == False
+-- 'hasLeftVertex' x (leftVertex x)   == True
+-- 'hasRightVertex' x (leftVertex x)  == False
+-- @
+leftVertex :: a -> AdjacencyMap a b
+leftVertex x = BAM (Map.singleton x Set.empty) Map.empty
+
+-- | Constructs the bipartite graph comprising /a single isolated vertex/ in
+-- the right part.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (rightVertex x)  == Map.'Map.empty'
+-- 'rightAdjacencyMap' (rightVertex x) == Map.'Map.singleton' x Set.'Set.empty'
+-- 'hasEdge' x y (rightVertex y)       == False
+-- 'hasLeftVertex' x (rightVertex x)   == False
+-- 'hasRightVertex' x (rightVertex x)  == True
+-- @
+rightVertex :: b -> AdjacencyMap a b
+rightVertex y = BAM Map.empty (Map.singleton y Set.empty)
+
+-- | Constructs the bipartite graph comprising /a single isolated vertex/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- vertex (Left x)                == 'leftVertex' x
+-- vertex (Right x)               == 'rightVertex' x
+-- 'hasEdge' x y (vertex (Left x))  == False
+-- 'hasEdge' x y (vertex (Right y)) == False
+-- @
+vertex :: Either a b -> AdjacencyMap a b
+vertex (Left x)  = leftVertex x
+vertex (Right y) = rightVertex y
+
+-- | Constructs the bipartite graph comprising /a single edge/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'leftAdjacencyMap' (edge x y)  == Map.'Map.singleton' x (Set.'Set.singleton' y)
+-- 'rightAdjacencyMap' (edge x y) == Map.'Map.singleton' y (Set.'Set.singleton' x)
+-- 'hasEdge' x y (edge x y)       == True
+-- 'hasEdge' y x (edge x y)       == (x == y)
```

Comment:
I find this example a bit difficult to understand. Perhaps, it's worth splitting it into two concrete ones, say `x=y=1`, and `x=1` and `y=2`.